text/trace/0220-messaging-semantic-conventions-span-structure.md

```diff
@@ -0,0 +1,639 @@
+end-to-end latency of processing a message. If intermediaries are not
+instrumented, this shouldn't impact the correlation of producer and consumer
+stages. If, on the other hand, intermediaries are instrumented, spans from
+intermediary instrumentation should seamlessly integrate with producer and
+consumer instrumentation. This integration should not require any changes in
+producer or consumer instrumentation, and it should not cause any changes to
+the relationships of producer and consumer spans. Furthermore, it should be
+possible to provide tracing instrumentation as an out-of-the-box experience
+from messaging SDKs, without requiring any additional custom instrumentation
+from the user.
+
+This OTEP aims at proposing consistent guidelines for creating spans that model
+the stages of the messaging flow, and correlating those in a way so that the
+requirements sketched above can be met in a consistent way across messaging
+scenarios and different messaging systems.
+
+## Terminology
+
+For terms used in this document, refer to [OTEP 173](#0173-messaging-semantic-conventions.md#terminology).
+
+## Stages of producing and consuming messages
+
+Producing and consuming a message involves five stages:
+
+```mermaid
+flowchart LR;
+  subgraph PRODUCER
+  direction TB
+  CR[Create] --> PU[Publish]
+  end
+  subgraph INTERMEDIARY
+  direction TB
+  end
+  subgraph CONSUMER
+  direction TB
+  RE[Receive] --> PR[Process]
+  PR --> SE[Settle]
+  end
+  PU --> INTERMEDIARY
+  INTERMEDIARY --> RE
+  SE -..-> INTERMEDIARY
+```
+
+1. The producer creates a message.
+2. The producer publishes the message to an intermediary.
+3. The consumer receives the message from an intermediary.
+4. The consumer processes the message.
+5. The consumer settles the message by notifying the intermediary that the
+   message was processed. In some cases (fire-and-forget scenarios, or when
+   settlement happens on the broker), the settlement stage does not exist.
+
+The semantic conventions described below define how to model those stages with
+spans.
+
+## Trace structure
+
+### Producer
+
+Producers are responsible for injecting a creation context into a message.
+Subsequent consumers will use this context to link consumer traces to producer
+traces. Ideally, each message gets a unique and distinct creation context
+assigned. However, as a context must refer to a span this would require the
+creation of a distinct span for each message, which is not feasible in all
+scenarios. In certain batching scenarios where many messages are created and
+published in large batches, creating a span for every single message would
+obfuscate traces and is not desirable. Thus instrumentation libraries and
+auto-instrumentation should default to creating a unique and distinct context
+per message, but may support configuration or other ways to change this default
+behavior.
+
+For each producer scenario, a "Publish" span needs to be created. This span
```

Comment:
under which trace id?

Reply:
The assumption here is, that spans are created as [defined by the specification](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#determining-the-parent-span-from-a-context): use the currently active span as parent (using its trace id), or otherwise start a new trace (with a new trace id).

Reply:
I guess this refers to the relationship between "Publish" and "Create" spans specifically?

If so, I tried to clarify [here](https://github.com/open-telemetry/oteps/pull/220/files#r1172940271):
* "Create" spans can be children of "Publish" spans. In this they are part of the same trace.
* "Create" spans can be part of the same trace as "Publish" spans without being children of the "Publish" operation. In this case "Publish" spans should link to "Create" spans.
* "Create" and "Publish" spans belong to different traces. In this case "Publish" spans should link to "Create" spans.